toolchain/base/int_store_test.cpp

```diff
@@ -0,0 +1,150 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/base/int_store.h"
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <limits>
+
+#include "toolchain/base/value_ids.h"
+
+namespace Carbon::Testing {
+
+struct IntStoreTestPeer {
+  static constexpr int MinAPWidth = IntStore::MinAPWidth;
+};
+
+namespace {
+
+using ::testing::Eq;
+using ::testing::Not;
+
+static constexpr int MinAPWidth = IntStoreTestPeer::MinAPWidth;
+
+TEST(IntStore, Basic) {
+  IntStore ints;
+  IntId id1 = ints.Add(1);
+  IntId id2 = ints.Add(2);
+  IntId id3 = ints.Add(999'999'999'999);
+
+  ASSERT_TRUE(id1.is_valid());
+  ASSERT_TRUE(id2.is_valid());
+  ASSERT_TRUE(id3.is_valid());
+  EXPECT_THAT(id1, Not(Eq(id2)));
+  EXPECT_THAT(id1, Not(Eq(id3)));
+  EXPECT_THAT(id2, Not(Eq(id3)));
+
+  EXPECT_THAT(ints.Get(id1), Eq(1));
+  EXPECT_THAT(ints.Get(id2), Eq(2));
+  EXPECT_THAT(ints.Get(id3), Eq(999'999'999'999));
+}
+
+TEST(IntStore, APSigned) {
+  IntStore ints;
+
+  llvm::APInt one_ap(MinAPWidth, 1, /*isSigned=*/true);
```

Comment:
I don't know whether you want to spend time on this, but I wonder if you could do something like:

```
struct TestValue {
  llvm::StringLiteral label;
  llvm::APInt value;
  testing::Matcher<llvm::APInt> matcher;
  // Set after initialization.
  IntId id = IntId::Invalid;
}

auto entries[] = {
  {"one_ap", llvm::APInt(MinAPWidth, 1, /*isSigned=*/true), Eq<llvm::APInt>(1)},
  ...
}

for (auto [i, entry] : llvm::enumerate(entries)) {
  SCOPED_TRACE(entry.label);
  entry.id = ints.AddSigned(entry.value);
  ASSERT_TRUE(entry.id.is_valid());
  for (int j : llvm::seq(i)) {
    EXPECT_THAT(entry.id, Not(Eq(entry[j].id)));
  }
}

for (auto entry : entries) {
  SCOPED_TRACE(entry.label);
  EXPECT_THAT(ints.Get(entry.id), entry.matcher);
}
```

Though not sure whether you like this idea or not -- do what you want with it, I was just thinking about it while trying to figure out which entries were being validated because "biggest_small_ap" has no correspondence with "ids[...]" for me.

Reply:
Looks good, thanks!